apollo-federation/src/sources/connect/json_selection/methods.rs

```diff
@@ -0,0 +1,2178 @@
+            })),
+            (
+                Some(json!({
+                    "names": ["Alice", "Bob", "Charlie"],
+                    "hobbies": ["reading", "fishing", "painting"],
+                })),
+                vec![],
+            ),
+        );
+    }
+
+    #[test]
+    fn test_typeof_method() {
+        fn check(selection: &str, data: &JSON, expected_type: &str) {
+            assert_eq!(
+                selection!(selection).apply_to(data),
+                (Some(json!(expected_type)), vec![]),
+            );
+        }
+
+        check("$->typeof", &json!(null), "null");
+        check("$->typeof", &json!(true), "boolean");
+        check("@->typeof", &json!(false), "boolean");
+        check("$->typeof", &json!(123), "number");
+        check("$->typeof", &json!(123.45), "number");
+        check("$->typeof", &json!("hello"), "string");
+        check("$->typeof", &json!([1, 2, 3]), "array");
+        check("$->typeof", &json!({ "key": "value" }), "object");
+    }
+
+    #[test]
+    fn test_map_method() {
+        assert_eq!(
+            selection!("$->map(@->add(10))").apply_to(&json!([1, 2, 3])),
+            (Some(json!(vec![11, 12, 13])), vec![]),
+        );
+
+        assert_eq!(
+            selection!("messages->map(@.role)").apply_to(&json!({
+                "messages": [
+                    { "role": "admin" },
+                    { "role": "user" },
+                    { "role": "guest" },
+                ],
+            })),
+            (Some(json!(["admin", "user", "guest"])), vec![]),
+        );
+
+        assert_eq!(
+            selection!("messages->map(@.roles)").apply_to(&json!({
+                "messages": [
+                    { "roles": ["admin"] },
+                    { "roles": ["user", "guest"] },
+                ],
+            })),
+            (Some(json!([["admin"], ["user", "guest"]])), vec![]),
+        );
+
+        assert_eq!(
+            selection!("values->map(@->typeof)").apply_to(&json!({
+                "values": [1, 2.5, "hello", true, null, [], {}],
+            })),
+            (
+                Some(json!([
+                    "number", "number", "string", "boolean", "null", "array", "object"
+                ])),
+                vec![],
+            ),
+        );
+
+        assert_eq!(
+            selection!("singleValue->map(@->mul(10))").apply_to(&json!({
+                "singleValue": 123,
+            })),
+            (Some(json!(1230)), vec![]),
+        );
+    }
+
+    #[test]
+    fn test_missing_method() {
+        assert_eq!(
+            selection!("nested.path->bogus").apply_to(&json!({
+                "nested": {
+                    "path": 123,
+                },
+            })),
+            (
+                None,
+                vec![ApplyToError::from_json(&json!({
+                    "message": "Method ->bogus not found",
+                    "path": ["nested", "path"],
+                }))],
+            ),
+        );
```

Comment:
We'll need to catch bogus methods (and invocations with the wrong parameters) at validation, _without_ running `apply_to`. Handling this at parse time is maybe the solution? Then we'd know they're correct at runtime, which might speed up `apply_to` a bit for selections we've already checked on router startup?

Reply:
I agree checking method names at parse time is reasonable: https://github.com/apollographql/router/pull/5762#discussion_r1715654729

However, our system for reporting parse errors is not great right now (no descriptive error messages or location information). That should be changing soon, but that's why I hesitated to make this change already.

Reply:
We do need to validate method names before handing this off to users, but it doesn't need to block this PR. Captured in https://apollographql.atlassian.net/browse/CNN-415.